src/Control/Concurrent/Execute.hs

```diff
@@ -56,26 +72,31 @@ instance Show ExecuteException where
 
 runActions :: Int -- ^ threads
            -> Bool -- ^ keep going after one task has failed
-           -> Bool -- ^ run final actions concurrently?
            -> [Action]
            -> (TVar Int -> IO ()) -- ^ progress updated
            -> IO [SomeException]
-runActions threads keepGoing concurrentFinal actions0 withProgress = do
+runActions threads keepGoing actions0 withProgress = do
     es <- ExecuteState
-        <$> newTVarIO actions0
+        <$> newTVarIO (sortActions actions0)
         <*> newTVarIO []
         <*> newTVarIO Set.empty
         <*> newTVarIO 0
-        <*> (if concurrentFinal
-                then pure Nothing
-                else Just <$> atomically (newTMVar ()))
         <*> pure keepGoing
     _ <- async $ withProgress $ esCompleted es
     if threads <= 1
         then runActions' es
         else replicateConcurrently_ threads $ runActions' es
     readTVarIO $ esExceptions es
 
+-- | Sort actions such that those that can't be run concurrently are at
+-- the end.
+sortActions :: [Action] -> [Action]
+sortActions = sortBy (compareConcurrency `on` actionConcurrency)
+  where
+    compareConcurrency ConcurrencyAllowed ConcurrencyDisallowed = LT
```

Comment:
What about `deriving Ord` on the `Concurrency` type?

Reply:
I considered that, but it would be easy for someone to add a constructor and not realize its impact on the usage of Ord.  I've added a note about it.